hastexo/jobs.py

```diff
@@ -91,10 +94,117 @@ def suspend_stack(self, stack):
                               "with status [%s]." % (stack.name, heat_status))
 
             # Schedule for retry, if it makes sense to do so
-            if (heat_status != DELETE_STATE and
+            if (heat_status != DELETED_STATE and
                     'FAILED' not in heat_status):
                 stack.status = SUSPEND_RETRY_STATE
             else:
                 stack.status = heat_status
 
             stack.save()
+
+
+class UndertakerJob(object):
+    """
+    Deletes old stacks.
+
+    """
+    settings = {}
+    stdout = None
+
+    def __init__(self, settings, stdout):
+        self.settings = settings
+        self.stdout = stdout
+
+    def run(self):
+        """
+        Delete stacks.
+
+        """
+        age = self.settings.get("delete_age", 14)
+        if not age:
+            return
+
+        timedelta = timezone.timedelta(days=age)
+        cutoff = timezone.now() - timedelta
+        dont_delete = [DELETE_STATE,
+                       DELETED_STATE,
+                       DELETE_IN_PROGRESS_STATE]
+
+        # Get stacks to delete
+        with transaction.atomic():
+            stacks = Stack.objects.select_for_update().filter(
+                suspend_timestamp__isnull=False
+            ).filter(
+                suspend_timestamp__lt=cutoff
+            ).exclude(
+                status__in=dont_delete
+            ).order_by('suspend_timestamp')
+
+            for stack in stacks:
+                stack.status = DELETE_STATE
+                stack.save()
+
+        # Delete them
+        for stack in stacks:
+            self.delete_stack(stack)
+
+    def get_heat_client(self, configuration):
+        return HeatWrapper(**configuration).get_client()
+
+    def delete_stack(self, stack):
+        """
+        Delete the stack.
+
+        """
+        configuration = get_xblock_configuration(self.settings, stack.provider)
+        timeouts = configuration.get('task_timeouts', {})
+        sleep = timeouts.get('sleep', 10)
+        retries = timeouts.get('retries', 90)
+        heat_client = self.get_heat_client(configuration)
+
+        self.stdout.write("Initializing stack [%s] deletion." % stack.name)
+
+        try:
+            heat_stack = heat_client.stacks.get(stack_id=stack.name)
+        except HTTPNotFound:
+            stack.status = DELETED_STATE
+            stack.save()
+        else:
+            self.stdout.write("Deleting stack [%s]." % (stack.name))
+
+            # Delete stack
+            heat_client.stacks.delete(stack_id=stack.name)
+
+            # Save status
+            stack.status = DELETE_IN_PROGRESS_STATE
+            stack.save()
+
+            # Wait until delete finishes.
+            retry = 0
+            while ('FAILED' not in stack.status and
```

Comment:
So if I read this correctly, this would retry for as long as Heat reports the stack as `DELETE_IN_PROGRESS`, and succeed if Heat either reports `DELETE_COMPLETE`, or issues a 404. However, it would stop if Heat reported `DELETE_FAILED`. Shouldn't it retry then, as well? (Consider that stack deletion can take several minutes, during which a compute node might die, or the Heat API may suffer an interruption, or `heat-engine` could be restarted due to maintenance.)